Stylo: replace product={gecko,servo} with engine={gecko,servo-2013,servo-2020}

SimonSapin opened the pull request:
Renaming the variable helped make sure I looked at every use.

Also auto-generate relevant parts of `CSSStyleDeclaration.webidl`, to make libscript compile accordingly.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/23856)
<!-- Reviewable:end -->

SimonSapin committed:
Stylo: replace product={gecko,servo} with engine={gecko,servo-2013,servo-2020}

Renaming the variable helped make sure I looked at every use.

highfive commented:
Heads up! This PR modifies the following files:
 * @wafflespeanut: python/servo/post_build_commands.py
 * @emilio: components/style/properties/properties.mako.rs, components/style/properties/longhands/inherited_table.mako.rs, components/style/properties/longhands/outline.mako.rs, components/style/properties/longhands/box.mako.rs, components/style/properties/shorthands/font.mako.rs and 44 more

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!
* These commits modify style and layout code, but no tests are modified. Please consider adding a test!

SimonSapin commented:
This still needs work in the script crate, to compile in `--with-layout-2020` mode.

emilio commented on components/style/properties/data.py:
```diff
@@ -615,9 +642,14 @@ def _add_logical_props(data, props):
 
 
+def remove(set_, item):
+    if item in set_:
```
I was intentionally using `remove()` so that getting it wrong errored loud, fwiw...

SimonSapin replied:
This is intended to catch a typo in the name, right?

My reasoning was that I don’t want to conditionally remove based on the engine, because we’d forget to edit this back when adding that property to the new layout engine.

I also tried not having a property exist at all before it is implemented in layout, but that’s already a lost cause: various parts of the style and script crate already refer to Rust APIs for specific APIs, so I used a pref to disable parsing instead. I’ll do the same here.

emilio commented on components/style/values/specified/box.rs:
```diff
@@ -37,6 +37,16 @@ fn moz_box_display_values_enabled(context: &ParserContext) -> bool {
+
+#[cfg(feature = "servo-layout-2020")]
+fn unimplemented_in_servo_2020(_context: &ParserContext) -> bool {
+    servo_config::prefs::pref_map().get("layout.2020.unimplemented").as_bool().unwrap_or(false)
```
This logic reads backwards... You only parse it if it's unimplemented?

I assume that this is what you want and that that pref would always be false, but why making it a pref at all? You could just stash `false` here and use `parse(condition = "not_servo_2020")` or such, which would read a bit nicer?

SimonSapin replied:
Oh yeah it’s totally backward :] Yes making it always false based on the build-time flag also works, but since there’s a pref for properties this felt more consistent.

SimonSapin replied:
Wait no, it’s not backward. The layout logic for this value is unimplemented. We only parse the keyword if the “enable unimplemented syntax” pref is on. I’ve renamed to `parse_unimplemented_in_servo_2020` to hopefully clarify.

This pref is basically never intended to be on, but prefs are what already had support in Mako template for disabling parsing of some properties while still generating a Rust type for their values.

emilio replied:
Right, my point is that it reads backwards. I understand what it's doing, it just reads "parse if unimplemented_in_servo_2020", which is confusing, I think. Prefixing it with `parse_` makes it a bit less confusing I guess.

emilio reviewed (commented):
The new variable name makes more sense, too :)

This looks generally good, let me know when it's ready for a final look.

SimonSapin committed:
Auto-generate CSSStyleDeclaration.webidl for CSS properties based on the style crate

SimonSapin commented:
Alright, this is ready for landing now.

r? @nox for the WebIDL-related commit
r? @emilio for the Stylo-related commit in case you have thoughts, but feel free to delegate to @nox.

jdm commented:
```
[0m[0m[1m[31merror:[0m failed to run custom build command for `script v0.0.1 (/repo/components/script)`

Caused by:
  process didn't exit successfully: `/repo/target/debug/build/script-c748da99a214968c/build-script-build` (exit code: 101)
--- stderr
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 2, kind: NotFound, message: "No such file or directory" }', src/libcore/result.rs:1051:5
stack backtrace:
   0: backtrace::backtrace::libunwind::trace
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.29/src/backtrace/libunwind.rs:88
   1: backtrace::backtrace::trace_unsynchronized
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.29/src/backtrace/mod.rs:66
   2: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:47
   3: std::sys_common::backtrace::print
             at src/libstd/sys_common/backtrace.rs:36
   4: std::panicking::default_hook::{{closure}}
             at src/libstd/panicking.rs:200
   5: std::panicking::default_hook
             at src/libstd/panicking.rs:214
   6: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:477
   7: std::panicking::continue_panic_fmt
             at src/libstd/panicking.rs:384
   8: rust_begin_unwind
             at src/libstd/panicking.rs:311
   9: core::panicking::panic_fmt
             at src/libcore/panicking.rs:85
  10: core::result::unwrap_failed
             at src/libcore/result.rs:1051
  11: core::result::Result<T,E>::unwrap
             at /rustc/273f42b5964c29dda2c5a349dd4655529767b07f/src/libcore/result.rs:852
  12: build_script_build::main
             at components/script/build.rs:21
  13: std::rt::lang_start::{{closure}}
             at /rustc/273f42b5964c29dda2c5a349dd4655529767b07f/src/libstd/rt.rs:64
  14: std::rt::lang_start_internal::{{closure}}
             at src/libstd/rt.rs:49
  15: std::panicking::try::do_call
             at src/libstd/panicking.rs:296
  16: __rust_maybe_catch_panic
             at src/libpanic_unwind/lib.rs:82
  17: std::panicking::try
             at src/libstd/panicking.rs:275
  18: std::panic::catch_unwind
             at src/libstd/panic.rs:394
  19: std::rt::lang_start_internal
             at src/libstd/rt.rs:48
  20: std::rt::lang_start
             at /rustc/273f42b5964c29dda2c5a349dd4655529767b07f/src/libstd/rt.rs:64
  21: main
  22: __libc_start_main
  23: _start
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

SimonSapin commented:
Added a line in `script/build.rs` to create the `$OUT_DIR/build/` directory before copying a file into it. In my local testing the directory was already there from an earlier build.

nox commented:
LGTM, feel free to r=me.

jdm commented:
@bors-servo r=nox

bors-servo commented:
:pushpin: Commit 2cc632e has been approved by `nox`

<!-- @bors-servo r=nox 2cc632e401a966f4a15cdd3ad1aec45ab16bc2d2 -->

bors-servo commented:
:hourglass: Testing commit 2cc632e401a966f4a15cdd3ad1aec45ab16bc2d2 with merge aa8b9fd347a515208feb7941a56691a49a756f96...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/bZIDPzdoQP2BCp6PdovAiQ)

SimonSapin committed:
Generate apis.html and css-properties.json for docs as part of crates’ build scripts

… rather than as an extra step after `cargo doc`.
This helps always using the correct set of CSS properties
(for layout 2013 v.s. 2020).

nox commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit e2f99a6 has been approved by `nox`

<!-- @bors-servo r=nox e2f99a6821551a4fb5fc71278bd97dea3006a9a5 -->

bors-servo commented:
:hourglass: Testing commit e2f99a6821551a4fb5fc71278bd97dea3006a9a5 with merge d0f3c73ce549607dd69052a9651b2bd0e1138630...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/BMzwn64ORQKJi1dYGf_Hxw)

CYBAI commented:
linux tidy complained 👀

```
Checking files for tidiness...

[94m./python/servo/post_build_commands.py[0m:[93m16[0m: [91mF401 'sys' imported but unused[0m
```

macOS and Windows failed on unit test of `properties_list_json` about reading the `css-properties.json`:

```
thread 'properties::scaffolding::properties_list_json' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 2, kind: NotFound, message: "The system cannot find the file specified." }', src\libcore\result.rs:1051:5
stack backtrace:
   0: std::sys_common::backtrace::_print
             at /rustc/273f42b5964c29dda2c5a349dd4655529767b07f\/src\libstd\sys_common\backtrace.rs:47
   1: std::panicking::default_hook::{{closure}}
             at /rustc/273f42b5964c29dda2c5a349dd4655529767b07f\/src\libstd\panicking.rs:200
   2: std::panicking::default_hook
             at /rustc/273f42b5964c29dda2c5a349dd4655529767b07f\/src\libstd\panicking.rs:211
   3: std::panicking::rust_panic_with_hook
             at /rustc/273f42b5964c29dda2c5a349dd4655529767b07f\/src\libstd\panicking.rs:477
   4: std::panicking::continue_panic_fmt
             at /rustc/273f42b5964c29dda2c5a349dd4655529767b07f\/src\libstd\panicking.rs:384
   5: std::panicking::rust_begin_panic
             at /rustc/273f42b5964c29dda2c5a349dd4655529767b07f\/src\libstd\panicking.rs:311
   6: core::panicking::panic_fmt
             at /rustc/273f42b5964c29dda2c5a349dd4655529767b07f\/src\libcore\panicking.rs:85
   7: core::result::unwrap_failed
             at /rustc/273f42b5964c29dda2c5a349dd4655529767b07f\/src\libcore\result.rs:1051
   8: core::result::Result<std::fs::File, std::io::error::Error>::unwrap<std::fs::File,std::io::error::Error>
             at /rustc/273f42b5964c29dda2c5a349dd4655529767b07f\src\libcore\result.rs:852
   9: style_tests::properties::scaffolding::properties_list_json
             at .\properties\scaffolding.rs:40
  10: style_tests::properties::scaffolding::properties_list_json::{{closure}}
             at .\properties\scaffolding.rs:12
  11: core::ops::function::FnOnce::call_once<closure,()>
             at /rustc/273f42b5964c29dda2c5a349dd4655529767b07f\src\libcore\ops\function.rs:231
  12: alloc::boxed::{{impl}}::call_once<(),FnOnce<()>>
             at /rustc/273f42b5964c29dda2c5a349dd4655529767b07f\src\liballoc\boxed.rs:766
  13: panic_unwind::__rust_maybe_catch_panic
             at /rustc/273f42b5964c29dda2c5a349dd4655529767b07f\/src\libpanic_unwind\lib.rs:82
  14: test::run_test::run_test_inner::{{closure}}
             at /rustc/273f42b5964c29dda2c5a349dd4655529767b07f\/src\libtest\lib.rs:1472
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

SimonSapin commented:
Removed the unused import, and removed the bit of unit test that runs a Python script that doesn’t exist anymore. (Kept the part where it check that the generated JSON file looks reasonable.)

@bors-servo r=nox

bors-servo commented:
:pushpin: Commit 0215d09 has been approved by `nox`

<!-- @bors-servo r=nox 0215d09ccbfed85c91605dedbb175dc9a96ba0ab -->

bors-servo commented:
:hourglass: Testing commit 0215d09ccbfed85c91605dedbb175dc9a96ba0ab with merge 8e7daa77b62a692beda57debf852d635cf24c49e...

bors-servo commented:
:sunny: Test successful - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/12992), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/13052), [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/E843dPfTRC-J6dXzzQKAyg)
Approved by: nox
Pushing 8e7daa77b62a692beda57debf852d635cf24c49e to master...
